### PR TITLE
ZED camra + imu for ROS2

### PR DESCRIPTION
**Pull request by rafal-gorecki**
Niewielkie zmiany w main.cpp, CMakeList i readme w stosunku do ROS1

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,73 @@
+cmake_minimum_required(VERSION 3.5)
+project(zed_cpu)
+
+set(CMAKE_CXX_STANDARD 14)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
```
Comment (KmakD) [inline CMakeLists.txt]:
```suggestion
cmake_minimum_required(VERSION 3.10.2)
project(zed_cpu)

set(CMAKE_CXX_STANDARD 17)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
```
Think we should go with c++17 for ROS2

File: include/zed_cpu/zed_cpu.hpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+#include <rclcpp/rclcpp.hpp>
+
+#include <sensor_msgs/msg/imu.hpp>
+
+#include <zed_lib/sensorcapture.hpp>
+#include <zed_lib/videocapture.hpp>
+
+namespace zed_cpu
+{
+class ZedCameraNode
```
Comment (KmakD) [inline include/zed_cpu/zed_cpu.hpp]:
```suggestion
namespace zed_cpu
{

class ZedCameraNode : public rclcpp::Node
```
For ROS 2 this approach is more common

File: include/zed_cpu/zed_cpu.hpp [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+{
+public:
+  ZedCameraNode(
+    const std::shared_ptr<rclcpp::Node> & nh,
+    const std::shared_ptr<image_transport::ImageTransport> & it);
```
Comment (KmakD) [inline include/zed_cpu/zed_cpu.hpp]:
```suggestion
  ZedCameraNode(const std::shared_ptr<image_transport::ImageTransport> & it);
```

File: package.xml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  <depend>compressed_image_transport</depend>
+  <depend>cv_bridge</depend>
+  <depend>image_transport</depend>
+  <depend>image_transport_plugins</depend>
+  <depend>libhidapi-dev</depend>
+  <depend>libopencv-dev</depend>
+  <depend>libusb-1.0-dev</depend>
+  <depend>sensor_msgs</depend>
+  <depend>std_msgs</depend>
+  <depend>rclcpp</depend>
```
Comment (KmakD) [inline package.xml]:
```suggestion
  <depend>rclcpp</depend>
  <depend>sensor_msgs</depend>
  <depend>std_msgs</depend>
```

File: package.xml [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+<package format="3">
+  <name>zed_cpu</name>
+  <version>1.0.0</version>
+  <description>Package for ZED cameras running on CPU</description>
+
+  <author email="rafal.gorecki@husarion.com">Rafał Górecki</author>
+  <maintainer email="support@husarion.com">Husarion</maintainer>
+  <url type="website">https://husarion.com/</url>
+  <url type="repository">https://github.com/husarion/zed_cpu</url>
+  <url type="bugtracker">https://github.com/husarion/zed_cpu/issues</url>
```
Comment (KmakD) [inline package.xml]:
```suggestion
  <maintainer email="support@husarion.com">Husarion</maintainer>

  <url type="website">https://husarion.com/</url>
  <url type="repository">https://github.com/husarion/zed_cpu</url>
  <url type="bugtracker">https://github.com/husarion/zed_cpu/issues</url>
```

File: src/main.cpp [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+{
+  rclcpp::init(argc, argv);
+  auto node = std::make_shared<rclcpp::Node>("zed_camera", "zed_camera");
+  auto it = std::make_shared<image_transport::ImageTransport>(node);
+
+  zed_cpu::ZedCameraNode zed_camera_node(node, it);
+  while (rclcpp::ok()) {
+    zed_camera_node.run();
+    rclcpp::spin_some(node);
+  }
```
Comment (KmakD) [inline src/main.cpp]:
```suggestion
  auto it = std::make_shared<image_transport::ImageTransport>(node);
  auto zed_camera_node = std::make_shared<zed_cpu::ZedCameraNode>(it);

  while (rclcpp::ok()) {
    zed_camera_node.run();
    rclcpp::spin_some(zed_camera_node);
  }
```

File: src/zed_cpu.cpp [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+#include <zed_lib/sensorcapture.hpp>
+#include <zed_lib/videocapture.hpp>
+
+namespace zed_cpu
+{
+
+ZedCameraNode::ZedCameraNode(
+  const std::shared_ptr<rclcpp::Node> & nh,
+  const std::shared_ptr<image_transport::ImageTransport> & it)
+: nh_(nh), it_(it)
```
Comment (KmakD) [inline src/zed_cpu.cpp]:
```suggestion
ZedCameraNode::ZedCameraNode(const std::shared_ptr<image_transport::ImageTransport> & it)
: Node("zed_camera_node"), it_(it)
```

File: src/zed_cpu.cpp [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+{
+  // ROS initialization
+  left_image_pub_ = it_->advertise("rgb/left_image", 1);
+  right_image_pub_ = it_->advertise("rgb/right_image", 1);
+  imu_pub_ = nh_->create_publisher<sensor_msgs::msg::Imu>("imu_data", 1);
```
Comment (KmakD) [inline src/zed_cpu.cpp]:
```suggestion
  imu_pub_ = this->create_publisher<sensor_msgs::msg::Imu>("imu_data", 1);
```

File: src/zed_cpu.cpp [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+  CameraInit();
+  SensorInit();
```
Comment (KmakD) [inline src/zed_cpu.cpp:28]:
```suggestion
  SensorInit();
  
  RCLCPP_INFO(get_logger(), "Node started");
```

File: src/zed_cpu.cpp [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+  // Initialize ZED camera
+  sl_oc::video::VideoParams params;
+  params.res = sl_oc::video::RESOLUTION::HD720;
+  params.fps = sl_oc::video::FPS::FPS_60;
+  params.verbose = sl_oc::VERBOSITY::ERROR;
+
+  // Create Video Capture
+  cap_ = std::make_unique<sl_oc::video::VideoCapture>(params);
+  if (!cap_->initializeVideo()) {
+    RCLCPP_ERROR(nh_->get_logger(), "Cannot open camera video capture");
```
Comment (KmakD) [inline src/zed_cpu.cpp]:
```suggestion
    RCLCPP_ERROR(get_logger(), "Cannot open camera video capture");
```
same for all logs

File: src/zed_cpu.cpp [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+
+void ZedCameraNode::PublishIMU()
+{
+  // Get IMU data with a timeout of 5 milliseconds
+  const sl_oc::sensors::data::Imu imu_data = sens_->getLastIMUData(5000);
+
+  if (imu_data.valid == sl_oc::sensors::data::Imu::NEW_VAL) {
+    // Create a sensor_msgs/Imu message
+    sensor_msgs::msg::Imu imu_msg;
+    imu_msg.header.stamp = nh_->get_clock()->now();
```
Comment (KmakD) [inline src/zed_cpu.cpp]:
```suggestion
    imu_msg.header.stamp = get_clock()->now();
```

File: README.md [hunk 12]
```diff
@@ -1,2 +1,70 @@
...
+ * Linux OS
+ * ROS Humble
+ * GCC (v7.5+)
+ * OpenCV (v3.4.0+)
+
+### Clone the repository
+
+    mkdir -p ros_ws/src
+    cd ros_ws
+    git clone https://github.com/husarion/zed-cpu.git src/zed-cpu
```
Comment (KmakD) [inline README.md]: add ```bash

File: README.md [hunk 13]
```diff
@@ -1,2 +1,70 @@
...
+
+    `sudo apt install libusb-1.0-0-dev libhidapi-libusb0 libhidapi-dev`
+
+* Install OpenCV:
+
+    `sudo apt install libopencv-dev libopencv-viz-dev`
+
+* Install ROS dependencies:
+
+    `rosdep install --from-paths src --ignore-src -r -y`
```
Comment (KmakD) [inline README.md]: add ``` bash

**Review by KmakD — CHANGES_REQUESTED** [review]

File: README.md (hunk 12 above)
Comment (rafal-gorecki) [inline README.md:23]:
```suggestion
```bash
mkdir -p ros_ws/src
cd ros_ws
git clone https://github.com/husarion/zed-cpu.git src/zed-cpu
```
```

File: src/zed_cpu.cpp [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+  // ROS initialization
+  left_image_pub_ = image_transport::create_publisher(this, "rgb/left_image");
+  right_image_pub_ = image_transport::create_publisher(this, "rgb/right_image");
+
+  imu_pub_ = create_publisher<sensor_msgs::msg::Imu>("imu_data", 1);
+
+  CameraInit();
+  SensorInit();
+
+  RCLCPP_INFO(get_logger(), "Node started");
```
Comment (KmakD) [inline src/zed_cpu.cpp]:
```suggestion
  imu_pub_ = this->create_publisher<sensor_msgs::msg::Imu>("imu_data", 1);

  CameraInit();
  SensorInit();

  RCLCPP_INFO(this->get_logger(), "Node started");
```
It is generally advised to include "this"  when calling member functions of current instance. It improves code clarity, and helps avoiding ambiguity (eg. class having function with the same name). Sorry for confusion with `get_logger`.

File: src/zed_cpu.cpp [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+
+void ZedCameraNode::PublishIMU()
+{
+  // Get IMU data with a timeout of 5 milliseconds
+  const sl_oc::sensors::data::Imu imu_data = sens_->getLastIMUData(5000);
+
+  if (imu_data.valid == sl_oc::sensors::data::Imu::NEW_VAL) {
+    // Create a sensor_msgs/Imu message
+    sensor_msgs::msg::Imu imu_msg;
+    imu_msg.header.stamp = get_clock()->now();
```
Comment (KmakD) [inline src/zed_cpu.cpp]:
```suggestion
    imu_msg.header.stamp = this->get_clock()->now();
```

File: src/zed_cpu.cpp [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+
+#include <sensor_msgs/msg/imu.hpp>
+
+#include <zed_lib/sensorcapture.hpp>
+#include <zed_lib/videocapture.hpp>
+
+namespace zed_cpu
+{
+
+ZedCameraNode::ZedCameraNode() : Node("zed_camera", "zed_camera")
```
Comment (KmakD) [inline src/zed_cpu.cpp:19]: I'm wondering if it wouldn't be a better idea to add namespace in launch file, with launch argument with default value "zed_camera". Better for multiple cameras implementations.

File: include/zed_cpu/zed_cpu.hpp [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+
+private:
+  void CameraInit();
+  void SensorInit();
+  void PublishImages();
+  void PublishIMU();
+
+  rclcpp::Publisher<sensor_msgs::msg::Imu>::SharedPtr imu_pub_;
+  image_transport::Publisher left_image_pub_;
+  image_transport::Publisher right_image_pub_;
```
Comment (KmakD) [inline include/zed_cpu/zed_cpu.hpp]: I think this should be unique_ptr. Alternatively, you can create ImageTransport shared_ptr object something like:
```
# zed_cpu.hpp
std::shared_ptr<image_transport::ImageTransport> it_;

# zed_cpu.cpp (constructor)
it_ = std::make_shared<image_transport::ImageTransport>(shared_from_this());
```

**Review by KmakD — CHANGES_REQUESTED** [review]

File: src/zed_cpu.cpp (hunk 16 above)
Comment (rafal-gorecki) [reply src/zed_cpu.cpp:19]: User still can use `namespace` argumet in launch file to override this

File: include/zed_cpu/zed_cpu.hpp (hunk 17 above)
Comment (rafal-gorecki) [reply include/zed_cpu/zed_cpu.hpp]: I tried this and it caused error: terminate called after throwing an instance of 'std::bad_weak_ptr' in this line:
`it_ = std::make_shared<image_transport::ImageTransport>(shared_from_this());`

In other hand using unique_ptr comes down to using the following command:
`left_image_pub_ = std::make_unique<image_transport::Publisher>(rclcpp::Node*, const string&, image_transport::PubLoaderPtr, rmw_qos_profile_t);` which arguments I don't understand and seems complicated.

**Pull request merged**